### DISCUSSION TITLE
Technology stack (Backend)

### DISCUSSION THREAD

**Discussion by steffen** [rarely-org/rarely-mobile#6] [General]
### Context

We aim to build a platform for allowing 1+ billion differently abled people to share their ideas and lived experiences and improve life.
User experience wise the platform is expected to be similar to social networks such as Facebook, Instagram, TikTok.

#### Firebase

[Firebase](https://firebase.google.com/) from Google is being considered.
It offers a single integrated solution for [authentication](https://firebase.google.com/products/auth) and [database](https://firebase.google.com/products/firestore).
On top of that it also offers [feature toggle functionality](https://firebase.google.com/products/remote-config) and a [messaging service](https://firebase.google.com/products/cloud-messaging).

#### Next.js/NextAuth/Prisma/PlanetScale

An alternative would be building our on backend with:
- [Next.js](https://nextjs.org/docs/api-routes/introduction)
- [NextAuth.js](https://next-auth.js.org/)
- [Prisma](https://www.prisma.io/)
- [PlanetScale](https://planetscale.com/)

Any thoughts?

**Comment by NVerban** [comment]
Next.js is more suitable for frontend web app, API endpoints in NextJS leaves a bit to be desired and may be limiting for a API app. Have you considered NestJS? https://nestjs.com

**Reply by steffen** [reply]
👋 Hi @NVerban, thank you for bringing up NestJS! I have not come across it yet to be honest.

You raise a good point: I'm not sure either if Next.js is the best choice for an API-only app.

Do you know how easy it is to integrate social logins such as Apple, Google, maybe Facebook/Meta, with it?
This is something that [NextAuth.js](https://next-auth.js.org/) seems to considerably help with (but then I'm not sure how helpful it is for an API-only app with a React Native mobile frontend).

For transparency, I took a brief look at [this comparison article](https://nodesource.com/blog/next-nuxt-nest/). NestJS seems to "win" in terms of performance.
Also, I haven't used annotation-driven frameworks yet, so that is an interesting aspect of NextJS and I assume a differentiator to Next.js?

**Reply by NVerban** [reply]
👋 Hi @steffen the main reason why I recommend NestJS is not actually primarily about performance (although it is very good), but rather that it is a fantastic framework that follows really good design patterns. But saying that, a big thing to consider is the team that will be working on this project and their experience. NestJS has a bit of a learning curve which probably would slow progress significantly initially if no one on the core team has experience with it (like any framework really). It’s a good idea to consider if you want to get this platform up and running quickly and as a community project have a lower barrier of entry for new comers to be able to contribute. IMO opting for a good framework can help keep things consistent, especially if contributors don’t have experience working together, on the other hand might be too steep of a learning curve for them to actually contribute much initially.

Implementing auth including SSO/OAUTH etc is not too difficult, NestJS can be setup with Express or Fastify and is really just extending these in a framework with some sensible conventions/patterns. So if you know how to implement Auth in Express for example, say with PassportJS, then it should be easy enough to implement in NestJS.

IMO, without a framework like NestJS, Express API apps can too easily become a big mess. Without a framework like NextJS, React apps can too easily become a big mess! And even with these frameworks, things can still become a big mess 😂 although framework conventions certainly help! The way I think about it, NextJS is best suited to frontend client apps (a React Framework). NestJS is for backend (a NodeJS framework). But either can do both. NestJS can be a full stack app and serve views / server side rendered templates, NextJS can run serverless or as standalone NodeJS app and serve API endpoints. But both seem to fall a bit short respectively as full stack frameworks.

IMO A winning combination is:

- Typescript FTW!
- NestJS API Backend + ExpressJS + TypeORM + PostgreSQL (or Prisma)
- NextJS frontend (if building a web client) **OR**
- React Native (if building native client)

Or if you just want to get a native mobile app app and running quickly and have simplest of stacks, just:

- Firebase (or equivalent)
- React Native

Keep it simple! And opt for what you have experience with. When considering shiny new tools for a team, I usually follow 80/20 rule. 80% of stack should be very familiar for the core team, and 20% shiny new things they want to try. This helps to ensure that development / scope / roadmap etc can be more predictable.

Hope this helps!

**Reply by steffen** [reply]
👋 Hi @NVerban, thank you so much for your very thoughtful response! I agree with everything and I like your 80/20 rule!

I love to go for simple solutions and I like the Firebase concept the most, which to my understanding allows authentication and access directly from the React Native client app without the need of a backend.
I've heard voices though that are very concerned about Firebase's Google connection and the implication on privacy.
There seem to exist alternatives for it, such as [Alby](https://ably.com/compare/ably-vs-firebase) and [Back4App](https://blog.back4app.com/best-firebase-alternatives/), which where just my first search results and that I haven't heard about before.

_IF_ we go with a backend, I tend to go with:
- **Typescript** - I believe strong types and explicit interfaces help community contributors to understand the codebase better and make safe contributions
- **TypeORM or Prisma** - I have a Ruby on Rails background and love ActiveRecord. [This article](https://www.prisma.io/docs/concepts/overview/prisma-in-your-stack/is-prisma-an-orm) seems to be a good article on comparing TypeORM and Prisma (with a potential bias towards Prisma as it is in their documentation 😊) which I need to read.
- **PostgreSQL or MySQL** - I have been hearing a lot about [PlanetScale](https://planetscale.com/) lately, which could be the 20% of shiny new things if we build our own backend. ☺️ (There is [this blog post](https://leerob.io/blog/mysql-planetscale) about migrating from Firebase to MySQL which probably has some good nuggets, but which I didn't read completely yet.)

There is no core team yet, hence right now a decision cannot be made based on the core team's preference and experience. 😊
Not sure if you will be able to join the [coding session this Saturday](https://www.linkedin.com/events/codeforagoodcausebycyprusjs6871136908521586688/about/), but maybe we can explore 1-2 solutions in that session.

**Comment by steffen** [comment]
Thanks to @saitonakamura for mentioning [**Supabase**](https://supabase.com/) in [yesterday's sessions](https://github.com/rarely-org/rarely/issues/9#issuecomment-986045751). I have been looking at it for a bit.

My first impression is that it's amazing, if not perfect, for Rarely! ✨ 

- It includes [authentication](https://supabase.com/auth) via social logins and passwordless (magic link), see [docs](https://supabase.com/docs/guides/auth)
- It's built on top of Postgres, see [docs](https://supabase.com/docs/guides/database)
- The docs look great, e.g. [going into production checklist](https://supabase.com/docs/going-into-prod)
- It's [open source](https://github.com/supabase/supabase), with a [very active community](https://github.com/supabase/supabase/discussions), see [overall stats](https://supabase.com/company)
- It has [great examples](https://supabase.com/docs/guides/examples), the first being an [Expo todo list](https://github.com/supabase/supabase/tree/master/examples/expo-todo-list)
  - Example for [auth flow with expo](https://github.com/codingki/react-native-expo-template/tree/master/template-typescript-bottom-tabs-supabase-auth-flow)
- It includes [object storage](https://supabase.com/docs/guides/storage)
- It offers a SaaS and a free self-hosted version. The [pricing](https://supabase.com/pricing) for the SaaS version looks totally acceptable for this project.
- [Their dashboard](https://supabase.com/blog/2021/11/30/supabase-studio) can be used for local development and for self-hosting
- Looks like there will also be coming [great logging capabilities](https://supabase.com/blog/2021/12/02/supabase-acquires-logflare)

/cc @PatrickHeneise @NVerban
